rbac/management/role/relation_api_dual_write_handler.py

```diff
@@ -34,16 +38,116 @@ class DualWriteException(Exception):
     pass
 
 
+class ReplicationEventType(str, Enum):
+    """Replication event type."""
+
+    CREATE_SYSTEM_ROLE = "create_system_role"
+    UPDATE_SYSTEM_ROLE = "update_system_role"
+    DELETE_SYSTEM_ROLE = "delete_system_role"
+    CREATE_CUSTOM_ROLE = "create_custom_role"
+    UPDATE_CUSTOM_ROLE = "update_custom_role"
+    DELETE_CUSTOM_ROLE = "delete_custom_role"
+    ASSIGN_ROLE = "assign_role"
+    UNASSIGN_ROLE = "unassign_role"
+    CREATE_GROUP = "create_group"
+    UPDATE_GROUP = "update_group"
+    DELETE_GROUP = "delete_group"
+
+
+class ReplicationEvent:
+    """What tuples changes to replicate."""
+
+    type: ReplicationEventType
+    partition_key: str
+    add: list[common_pb2.Relationship]
+    remove: list[common_pb2.Relationship]
+
+    def __init__(
+        self,
+        type: ReplicationEventType,
+        partition_key: str,
+        add: list[common_pb2.Relationship] = [],
+        remove: list[common_pb2.Relationship] = [],
+    ):
+        """Initialize ReplicationEvent."""
+        self.partition_key = partition_key
+        self.type = type
+        self.add = add
+        self.remove = remove
+
+
+class RelationReplicator(ABC):
+    """Type responsible for replicating relations to Kessel Relations."""
+
+    @abstractmethod
+    def replicate(self, event: ReplicationEvent):
+        """Replicate the given event to Kessel Relations."""
+        pass
+
+
+class OutboxReplicator(RelationReplicator):
+    """Replicates relations via the outbox table."""
+
+    def __init__(self, role):
+        """Initialize OutboxReplicater."""
+        self.role = role
```

Comment:
or maybe it will work with any object which has uuid and name or those ones we are logging. So probably we just need to rename it to something general like `record` and ensure that object has attributes which we are going to log/use.

Reply:
Yeah I think we should do this, I just didn't take on this refactoring with this PR. Do you want to take on this adjustment in your group pr?

Reply:
yes I can do it 👍